File: DSA/Hima-Swarupa-Padala/day18q2.py
class Node:
    def __init__(self, freq, char=None):
        self.freq = freq
        self.char = char
        self.left = None
        self.right = None

def decode_huffman(root, s):
    decoded_text = ""
    current = root
    for bit in s:
        if bit == '0':
            current = current.left
        else:
            current = current.right
        if current.char:
            decoded_text += current.char
            current = root
    return decoded_text

root = Node(None)
root.left = Node(None)
root.right = Node(3, 'A')
root.left.left = Node(1, 'B')
root.left.right = Node(1, 'C')

huffman_tree = root
binary_string = "1001011"

decoded_text = decode_huffman(huffman_tree, binary_string)
print(decoded_text)  

File: DSA/Shruti-Pandey/Day_18_q2.java
import java.util.*;

class HuffmanNode {
    char data;
    HuffmanNode left, right;

    public HuffmanNode() {
        left = right = null;
    }
}

public class HuffmanDecoder {
    public static String decode(HuffmanNode root, String binaryString) {
        StringBuilder decodedText = new StringBuilder();
        HuffmanNode current = root;

        for (char bit : binaryString.toCharArray()) {
            if (bit == '0') {
                current = current.left;
            } else {
                current = current.right;
            }

            if (current.left == null && current.right == null) {
                decodedText.append(current.data);
                current = root;
            }
        }

        return decodedText.toString();
    }

    public static void main(String[] args) {
        HuffmanNode root = new HuffmanNode();
        root.left = new HuffmanNode();
        root.right = new HuffmanNode();
        root.left.data = 'B';
        root.right.left = new HuffmanNode();
        root.right.right = new HuffmanNode();
        root.right.left.data = 'C';
        root.right.right.data = 'A';

        String binaryString = "1001011";
        String decodedText = decode(root, binaryString);
        System.out.println("Decoded text: " + decodedText);
    }
}
